src/bindings/swift/gen_swift.rs

```diff
@@ -0,0 +1,498 @@
+        self.offset = 0
+    }
+
+    func writeInt<T: FixedWidthInteger>(_ value: T) {
+        var value = value.bigEndian
+        let _ = withUnsafeBytes(of: &value, { bytes.append(contentsOf: $0) })
+    }
+
+    @inlinable
+    func writeUInt8(_ value: UInt8) {
+        writeInt(value)
+    }
+
+    @inlinable
+    func writeUInt32(_ value: UInt32) {
+        writeInt(value)
+    }
+
+    @inlinable
+    func writeUInt64(_ value: UInt64) {
+        writeInt(value)
+    }
+}
+
+protocol Lowerable {
+    func lower(into buf: DataWriter)
+}
+
+protocol Liftable {
+    static func lift(from buf: DataReader) -> Self
+}
+
+func lowerIntoRustBuffer<T: Lowerable>(_ value: T) -> RustBuffer {
+    let writer = DataWriter()
+    value.lower(into: writer)
+    return RustBuffer(bytes: writer.bytes)
+}
+
+func liftFromRustBuffer<T: Liftable>(_ buf: RustBuffer) -> T {
+    let reader = DataReader(data: Data(rustBuffer: buf))
+    let value = T.lift(from: reader)
+    if reader.hasRemaining() {
+        fatalError()
+    }
+    buf.deallocate()
+    return value
+}
+
+extension Bool : Lowerable, Liftable {
+    func lower() -> UInt8 {
+        return self ? 1 : 0
+    }
+
+    func lower(into buf: DataWriter) {
+        buf.writeUInt8(self.lower())
+    }
+
+    static func lift(_ v: UInt8) -> Bool {
+        return v != 0
+    }
+
+    static func lift(from buf: DataReader) -> Bool {
+        return self.lift(buf.readUInt8())
+    }
+}
+
+extension UInt8: Lowerable, Liftable {
+    func lower() -> UInt8 {
+        return self
+    }
+
+    func lower(into buf: DataWriter) {
+        buf.writeUInt8(self.lower())
+    }
+
+    static func lift(_ v: UInt8) -> UInt8 {
+        return v
+    }
+
+    static func lift(from buf: DataReader) -> UInt8 {
+        return self.lift(buf.readUInt8())
+    }
+}
+
+extension UInt32: Lowerable, Liftable {
+    func lower() -> UInt32 {
+        return self
+    }
+
+    func lower(into buf: DataWriter) {
+        buf.writeUInt32(self.lower())
+    }
+
+    static func lift(_ v: UInt32) -> UInt32 {
+        return v
+    }
+
+    static func lift(from buf: DataReader) -> UInt32 {
+        return self.lift(buf.readUInt32())
+    }
+}
+
+extension UInt64: Lowerable, Liftable {
+    func lower() -> UInt64 {
+        return self
+    }
+
+    func lower(into buf: DataWriter) {
+        buf.writeUInt64(self.lower())
+    }
+
+    static func lift(_ v: UInt64) -> UInt64 {
+        return v
+    }
+
+    static func lift(from buf: DataReader) -> UInt64 {
+        return self.lift(buf.readUInt64())
+    }
+}
+
+extension Optional: Lowerable where Wrapped: Lowerable {
+    func lower() -> RustBuffer {
+        lowerIntoRustBuffer(self)
+    }
+
+    func lower(into buf: DataWriter) {
+        guard let value = self else {
+            buf.writeUInt8(0)
+            return
+        }
+        buf.writeUInt8(1)
+        value.lower(into: buf)
+    }
+}
+
+extension Optional: Liftable where Wrapped: Liftable {
```

Comment:
I don't have any deep swift knowledge, and this protocol/extension stuff feels surprisingly nice, and a good mental map back to how it's done on the Rust side. I like it!

Reply:
Yeah, this was super fun to write! I haven't worked much with Swift before, but its generics feel really familiar coming from Rust, which is exciting! 🦀